File: modules/core/app/model/exercise/Exercise.java
package model.exercise;

import java.nio.file.Path;
import java.util.Collections;
import java.util.List;

import javax.persistence.Column;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.Id;
import javax.persistence.MappedSuperclass;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.base.Splitter;

import io.ebean.Model;
import model.JsonReadable;
import play.twirl.api.Html;

@MappedSuperclass public abstract class Exercise extends Model implements JsonReadable {

  protected static final Splitter SPLITTER = Splitter.fixedLength(100).omitEmptyStrings();

  protected static final Splitter NEW_LINE_SPLITTER = Splitter.on("\n");

  public static final String SPLIT_CHAR = "#";

  @Id public int id;

  @Column @JsonProperty(required = true) public String title;

  @Column @JsonProperty(required = true) public String author;

  @Column(columnDefinition = "text") @JsonProperty(required = true) public String text;

  @Column @Enumerated(EnumType.STRING) public ExerciseState state;

  public Exercise(int theId) {
    id = theId;
  }

  @Override public int getId() {
    return id;
  }

  @JsonIgnore public List<? extends Tag> getTags() {
    return Collections.emptyList();
  }

  @JsonGetter("text") public List<String> getTextForJson() {
    return SPLITTER.splitToList(text);
  }

  public Html renderRest(scala.collection.immutable.List<scala.util.Try<Path>> fileResults) {
    return new Html("");
  }

  public Html renderEditRest(boolean isCreation) {
    return new Html("");
  }

  @Override public void saveInDB() {
    save();
  }

}

File: modules/core/app/model/exercise/Exercise.scala
package model.exercise

import java.nio.file.Path
import javax.persistence._

import com.fasterxml.jackson.annotation.{JsonGetter, JsonIgnore, JsonProperty}
import com.google.common.base.Splitter
import io.ebean.Model
import model.JsonReadable
import model.exercise.Exercise._
import play.twirl.api.Html
import scala.collection.JavaConverters._
import scala.util.Try

object Exercise {
  val SPLIT_CHAR = "#"

  val SPLITTER: Splitter = Splitter.fixedLength(100).omitEmptyStrings()

  val NEW_LINE_SPLITTER: Splitter = Splitter.on("\n")
}

@MappedSuperclass
abstract class Exercise extends Model with JsonReadable {

  @Id
  var id: Int

  @Column
  @JsonProperty(required = true)
  var title: String = _

  @Column
  @JsonProperty(required = true)
  var author: String = _

  @Column(columnDefinition = "text")
  @JsonProperty(required = true)
  var text: String = _

  @Column
  @Enumerated(EnumType.STRING)
  var state: ExerciseState = _


  override def getId: Int = id

  @JsonIgnore
  def getTags: java.util.List[Tag] = List.empty.asJava

  @JsonGetter("text")
  def getTextForJson: java.util.List[String] = SPLITTER.splitToList(text)

  def renderRest(fileResults: List[Try[Path]]) = new Html("")

  def renderEditRest(isCreation: Boolean) = new Html("")

  override def saveInDB(): Unit = save()

}
